docs/proposals/api-group.md

```diff
@@ -0,0 +1,135 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/api-group.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Supporting multiple API groups
+
+## Goal
+
+1. Breaking the monolithic v1 API into modular groups
+
+2. Supporting different versions in different groups
+
+3. Supporting identically named kinds to exist in different groups
+
+4. Enabling/disabling groups individually
+
+5. Allowing client to discover the API groups supported by the server
+
+6. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
+
+7. Allowing users to omit group name when using kubectl
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  For fast implementation, we will not add a new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+    * for api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
+
+    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
+
+    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
```

Comment:
It's important to remember that /osapi is not mounted on the same server as
Kube - by definition, it would be on a different server with a different
base URL.

> Could you develop your point? What do you mean by "arbitrary"? Does the
client know the group/version/kind of the object?

Jordan's question was specifically around how does a practical client deal
with a list of resources that come from different servers or groups.  We
have to have some "discovery" process whereby all groups are known to each
client (as @abonas mentioned, this is a traditional discovery pattern).
The outcome of that discovery is a mapping of (apiVersion, kind, group?) ->
client capable of dealing with the group (what we call client.RESTClient
today).  So a client that got back an `api.List` with tuples of
(apiVersion,kind,group) has to identify those clients.

Duplicates are possible - I think it means that clients have to allow users
to resolve duplicates at interaction time or preferences time.  If a client
adds a server that introduces a duplicate REST mapping for a particular
kind, how do they interact with that resource?  So conflict
detection/resolution has to be built into the RESTMapper that resolves
generic kind/version tuples.  Scheme is really the last part of the process
- we only call into a particular Scheme when all duplication behavior is
resolved.

The creation of a Scheme by introspection of a remote server's swagger
document also has to have conflict resolution.


On Tue, Aug 25, 2015 at 3:31 PM, Daniel Smith <notifications@github.com>
wrote:

> In docs/proposals/api-group.md
> <https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37901441>
> :
>
> > +
> > +  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
> > +
> > +  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
> > +
> > +## Server-side implementation
> > +
> > +1. resource handlers' URL:
> > +
> > +  We will force the URL to be in the form of prefix/group/version/...
> > +
> > +    * for api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
> > +
> > +    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
> > +
> > +    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
>
> This means the prefix is a variable, sometimes equals to "apis/" and
> sometimes equals to "", right?
>
> I would say it differently-- there's no such thing as prefix; we mount all
> apis under e.g. /apis, with special apis also mounted elsewhere for
> backwards compatibility (e.g. /api; /osapi).
>
> Also, I want to make the group of v1 APIs the empty string,
>
> It seems like this will be very confusing, but maybe you can make it work.
> I'm not sure. Perhaps we'll have to do this to not break old clients.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12951/files#r37901441>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift